src/transaction/output.rs

```diff
@@ -0,0 +1,469 @@
+            .iter()
+            .map(|out| RevealedOutputCommitment {
+                public_key: out.public_key,
+                revealed_commitment: out.random_commitment(&mut rng),
+            })
+            .take(self.outputs.len() - 1)
+            .collect();
+
+        // todo: replace fold() with sum() when supported in blstrs
+        let input_sum: Scalar = revealed_input_commitments
+            .iter()
+            .map(RevealedCommitment::blinding)
+            .fold(Scalar::zero(), |sum, x| sum + x);
+
+        // todo: replace fold() with sum() when supported in blstrs
+        let output_sum: Scalar = revealed_output_commitments
+            .iter()
+            .map(|r| r.revealed_commitment.blinding())
+            .fold(Scalar::zero(), |sum, x| sum + x);
+
+        let output_blinding_correction = input_sum - output_sum;
+
+        if let Some(last_output) = self.outputs.last() {
+            revealed_output_commitments.push(RevealedOutputCommitment {
+                public_key: last_output.public_key,
+                revealed_commitment: RevealedCommitment {
+                    value: last_output.amount,
+                    blinding: output_blinding_correction,
+                },
+            });
+        } else {
+            panic!("Expected at least one output")
+        }
+        revealed_output_commitments
+    }
+
+    fn output_range_proofs(
+        &self,
+        revealed_output_commitments: &[RevealedOutputCommitment],
+        mut rng: impl RngCore + CryptoRng,
+    ) -> Result<Vec<OutputProof>> {
+        let mut prover_ts = Transcript::new(MERLIN_TRANSCRIPT_LABEL);
+
+        let bp_gens = Self::bp_gens();
+
+        revealed_output_commitments
+            .iter()
+            .map(|c| {
+                let (range_proof, commitment) = RangeProof::prove_single_with_rng(
+                    &bp_gens,
+                    &Self::pc_gens(),
+                    &mut prover_ts,
+                    c.revealed_commitment.value,
+                    &c.revealed_commitment.blinding,
+                    RANGE_PROOF_BITS,
+                    &mut rng,
+                )?;
+
+                Ok(OutputProof {
+                    public_key: c.public_key,
+                    range_proof,
+                    commitment,
+                })
+            })
+            .collect::<Result<Vec<_>>>()
+    }
+}
+
+// note: used by both RevealedTransaction::sign and DbcTransaction::verify()
+//       which must match.
+fn gen_message_for_signing(
+    public_keys: &[PublicKey],
+    input_commitments: &[Commitment],
+    output_proofs: &[OutputProof],
+) -> Vec<u8> {
+    // Generate message to sign.
+    let mut msg: Vec<u8> = Default::default();
+    for pk in public_keys.iter() {
+        msg.extend(pk.to_bytes().as_ref());
+    }
+    for r in input_commitments.iter() {
+        msg.extend(r.to_bytes().as_ref());
+    }
+    for o in output_proofs.iter() {
+        msg.extend(o.to_bytes());
+    }
+    msg
```

Comment:
This is a paranoia thing but it's a good idea to put domain separators between the different field types to avoid situations where different messages may have the same byte encoding.

i.e. say for msg1, you had public_keys=[a,b,c] and commitments = [d]
and then msg2 where public_keys=[a,b] and commitments=[c,d]

If you look at the flattened byte encoding of the two messages, you'd see `<a><b><c><d>` for both messages.
Compare with domain seperators: 
msg1: `public_keys<a><b><c>commitments<d>`
msg2: `public_keys<a><b>commitments<c><d>`

```suggestion
    // Generate message to sign.
    let mut msg: Vec<u8> = Default::default();
    msg.extend("public_keys".as_bytes());
    for pk in public_keys.iter() {
        msg.extend(pk.to_bytes().as_ref());
    }
    msg.extend("commitments".as_bytes());
    for r in input_commitments.iter() {
        msg.extend(r.to_bytes().as_ref());
    }
    msg.extend("output_proofs".as_bytes());
    for o in output_proofs.iter() {
        msg.extend(o.to_bytes());
    }
    msg.extend("end".as_bytes());
    msg
```